Analysis of Fox2/3 KO Experiment
================================
## To Do

## Setup

```{r, message=FALSE}
knitr::opts_chunk$set(message=FALSE, warning =FALSE)
library('DEXSeq')
library('Rsubread')
library(tidyverse)
library(goseq)
setwd('~/Projects/Clients/Turnbull/CRISPR3_Take2/')
```

## Import data

```{r}
# Set up file locations and sample names
dirs = list.dirs('results/05_map2',recursive = FALSE)
# dirs
samps = gsub('results/05_map2/', '', dirs, fixed = TRUE)
samps

# Set up the Gene ID lookup table (csv was produced manually from the GTF file)
gene_id_lookup = read.csv('refs/unique_gene_ids.csv')
dim(gene_id_lookup)
# head(gene_id_lookup)
n_distinct(gene_id_lookup$gene_id)
n_distinct(gene_id_lookup$GeneID)
n_distinct(gene_id_lookup)
gene_id_v = gene_id_lookup$GeneID
names(gene_id_v) = gene_id_lookup$gene_id
(duplicate_geneIDs = (gene_id_lookup 
                     %>% count(GeneID) 
                     %>% filter(n > 1) 
                     %>% left_join(gene_id_lookup)))


# Pre-make the data frames to fill all the numbers in
exon_counts = matrix(NA, ncol = length(samps), nrow = 1520722)
colnames(exon_counts) = samps
# head(exon_counts)

# Define the gtf and genome files
gtf = '~/Disk2/CommonData/ReferenceGenomes/Mouse/GRCm39/ncbi-genomes-2022-03-04/GCF_000001635.27_GRCm39_genomic.gtf'
genome = '~/Disk2/CommonData/ReferenceGenomes/Mouse/GRCm39/ncbi-genomes-2022-03-04/GCF_000001635.27_GRCm39_genomic.fna'

# Bring in the sample data and clean it up
samdat = read.csv('data/sample_data/samdat.csv', row.names = 1)
samdat
samdat = (samdat
          %>% mutate(Treatment = factor(Treatment, levels = c('WT','KO')),
                     Differentiation = factor(Differentiation, 
                                              levels = c('Naive', 'Diff'))))
samps = rownames(samdat)
samps = gsub('-','_',samps,fixed = TRUE)
rownames(samdat) = samps
```

## Feature Counts

### Setup
```{r}
# Define a function to count features at the exon and gene level
call_fc = function(bamf, gtf, genome, useMF = FALSE){
    fc = Rsubread::featureCounts(bamf, 
                       annot.ext = gtf, isGTFAnnotationFile = TRUE,
                       GTF.featureType = 'exon', 
                       GTF.attrType = 'gene_id',
                       GTF.attrType.extra = 'exon_number', 
                       allowMultiOverlap = TRUE,
                       strandSpecific = 2,
                       genome = genome,
                       isPairedEnd = TRUE,
                       countReadPairs = FALSE,
                       requireBothEndsMapped = TRUE,
                       countChimericFragments = FALSE,
                       nthreads = 10,
                       verbose = TRUE,
                       useMetaFeatures = useMF)
    return(fc)
}
```

### Run on the first sample

I run this once and then keep it stored because it takes a long time
to run

```{r}
# # Call it all once to set up the structure
# samp = samps[1]
# dr = dirs[1]
# bamf = paste(dr, 'Aligned.sortedByCoord.out.bam', sep = '/')
# fc = call_fc(bamf, gtf, genome)
# fc_genes = call_fc(bamf, gtf, genome, TRUE)
# stats = (fc$stat
#          %>% data.frame()
#          %>% column_to_rownames('Status'))
# colnames(stats) = samp
# exon_counts[,samp] = fc$counts
# 
# fc_lst = list()
# fc_lst[[samp]] = fc
# fc_lst_gene = list()
# fc_lst_gene[[samp]] = fc_genes
```

### Run on the rest of the samples

I run this once and then keep it stored because it takes a long time
to run

```{r}
# # Call the same thing on all the mapping files
# 
# for (i in 2:length(dirs)){
#     samp = samps[i]
#     dr = dirs[i]
#     bamf = paste(dr, 'Aligned.sortedByCoord.out.bam', sep = '/')
#     fc = call_fc(bamf, gtf, genome)
#     exon_counts[,samp] = fc$counts
#     stats[fc$stat$Status,samp] = fc$stat$Aligned.sortedByCoord.out.bam
#     fc_lst[[samp]] = fc
#     fc_genes = call_fc(bamf, gtf, genome, useMF = TRUE)
#     fc_lst_gene[[samp]] = fc_genes
# }
# 
# # get gene lengths
# gene_lens = select(fc_genes$annotation, GeneID, Length)
# 
# save(list = c('exon_counts', 'stats', 'fc_lst', 'fc_lst_gene', 'gene_lens'), 
#      file = 'RData/feature_counts.RData')
load('RData/feature_counts.RData')
# head(exon_counts)
```


### Parse the feature counts

```{r}
# Get the gene and exon information for all the files
# chk = vector(length = length(fc_lst) -1)
# for (i in 2:length(fc_lst)){
#     chk[i-1] = all_equal(fc_lst[[1]]$annotation, fc_lst[[i]]$annotation)
# }
annot = fc_lst[[1]]$annotation
# head(annot)

# Make sure exon numbers are unique and in a format DEXSeq can use
annot = (annot
         %>% data.frame()
         %>% group_by(GeneID)
         %>% mutate(exon_unique = as.character(1:length(GeneID)))
         %>% data.frame())
# head(annot)
# head(rownames(exon_counts))

# Make sure sample names are correctly formatted
samps = colnames(exon_counts)
samps = gsub('-','_',samps, fixed = TRUE)
colnames(exon_counts) = samps
```

Check if the two pairs of genes with duplicate IDs have any read
counts associated with them

```{r}
# Do the genes with duplicate IDs have any counts associated with them?
rowSums(exon_counts[annot$GeneID %in% duplicate_geneIDs$gene_id,])

# They do. In some cases, it's a pretty big number
dup_cts = exon_counts[annot$GeneID %in% duplicate_geneIDs$gene_id,] 
rownames(dup_cts) = annot$GeneID[annot$GeneID %in% duplicate_geneIDs$gene_id]
dup_cts

annot %>% filter(GeneID %in% duplicate_geneIDs$gene_id)
```

These genes can't be merged because they are different, but they have the same
GeneID. They will be included in the DEXSeq and DESeq analyses, but excluded
from the goseq analysis.

## Differential Exon Usage

### Naive cells

#### DEXSeq

I run this once and then keep it stored because it takes a long time to run

```{r}
# Run DEXSeq on undifferentiated samples
# head(exon_counts)

samdat_undif = filter(samdat, Differentiation == 'Naive')
exon_ct_undif = exon_counts[,rownames(samdat_undif)]
samdat_undif
# head(exon_ct_undif)
# 
# dxd = DEXSeqDataSet(exon_ct_undif, samdat_undif,
#                     design = ~sample + exon + exon:Treatment,
#                     featureID = annot$exon_unique,
#                     groupID = annot$GeneID)
# dxd = estimateSizeFactors(dxd, locfunc = genefilter::shorth)
# dxd = estimateDispersions(dxd, BPPARAM = MulticoreParam(workers = 10))
# save(dxd, file = 'RData/dxd_undif.RData')
# plotDispEsts(dxd)
# dxd = testForDEU(dxd, BPPARAM = MulticoreParam(workers = 10))
# dxd = estimateExonFoldChanges(dxd, fitExpToVar = 'Treatment', 
#                               BPPARAM = MulticoreParam(workers = 10))
# save(dxd, file = 'RData/dxd_undif_deu.RData')
load('RData/dxd_undif_deu.RData')
res_orig_dex_undif = DEXSeqResults(dxd)

# Identify genes with significanly different (padj < 0.05) exon usage in at 
# least one exon
# head(res_orig_dex_undif)
dim(res_orig_dex_undif)

# Remove the exons with insufficient data to analyze
res_dex_undif = (res_orig_dex_undif
       %>% data.frame()
       %>% filter(!is.na(padj)))
dim(res_dex_undif)

sum(res_dex_undif$padj < 0.05)

sig_dex_undif = (res_dex_undif
                 %>% filter(padj < 0.05)
                 %>% select(-genomicData))
sig_10_dex_undif = (res_dex_undif
                    %>% filter(padj < 0.1)
                    %>% select(-genomicData))
# head(sig_dex_undif)
rownames(sig_dex_undif) = NULL
write.csv(sig_dex_undif, file = 'results/06_analysis/naive_deu_sig.csv')
write.csv(sig_10_dex_undif, file = 'results/06_analysis/naive_deu_0.1_sig.csv')
# write.csv(select(res_dex_undif, -genomicData), file = 'results/06_analysis/naive_deu_all.csv',
#             row.names = FALSE)
```

There are 1,520,722 exons analyzed. Of them, 1,125,713 had sufficient data for
analysis, and 6,520 had adjusted p-values less than 0.05. You can download the
significant results of DEXSeq analysis
[here](./results/06_analysis/naive_deu_sig.csv')


Pull out the genes that have at least one DEU exon:

```{r}
sig_genes_dex_undif = (res_dex_undif
           %>% filter(padj < 0.05)
           %>% select(groupID)
           %>% unique())$groupID
length(sig_genes_dex_undif)

# Check if any of the problem genes have DEU
any(sig_genes_dex_undif %in% duplicate_geneIDs$gene_id)
```

989 genes have at least one exon with DEU. None of the genes with duplicate IDs 
have DEU.

#### Goseq

```{r}
all_genes = unique(res_orig_dex_undif$groupID)
length(all_genes)
genes_dex_undif = unique(res_dex_undif$groupID)
length(genes_dex_undif)
```

Of the 40,840 genes in the data set, 19,028 had sufficient coverage for analysis
in at least one exon.

```{r, warning=FALSE, message=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_dex_undif = genes_dex_undif %in% sig_genes_dex_undif
deg_v_dex_undif = as.numeric(deg_v_dex_undif)
names(deg_v_dex_undif) = genes_dex_undif

# remove the duplicate genes
deg_v_dex_undif = deg_v_dex_undif[!(genes_dex_undif %in% duplicate_geneIDs$gene_id)]
names(deg_v_dex_undif) = gene_id_v[names(deg_v_dex_undif)]
# head(deg_v_dex_undif)
any(is.na(names(deg_v_dex_undif)))

# Get the bias and length estimations for goseq
np_dex_undif = nullp(deg_v_dex_undif, 'mm39', 'refGene')

# Run goseq
gs_dex_undif = goseq(np_dex_undif, 'mm39', 'refGene')

# head(gs_dex_undif)
dim(gs_dex_undif)
gs_fixed_dex_undif = (gs_dex_undif
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_dex_undif)

# Count the significant categories
sum(gs_fixed_dex_undif$padj < 0.05)

# Write the results to a file
write.csv(filter(gs_fixed_dex_undif, padj < 0.05), 
          file = 'results/06_analysis/goseq_deu_naive.csv')
write.csv(gs_fixed_dex_undif, file = 'results/06_analysis/goseq_deu_naive_all.csv')
```

The 989 genes with significantly different exon usage produced 
over-representation of DEU in 220 GO categories (out of a total of 21,389 
possible categories, after BH FDR adjustment). You can download the significant
results of the `goseq` analysis
[here](./results/06_analysis/goseq_deu_naive.csv), or the full results 
[here](./results/06_analysis/goseq_deu_naive_all.csv).


### Differentiated cells

#### DEXSeq

I run this once and then keep it stored because it takes a long time to run

```{r}
# # Run DEXSeq on differentiated samples
# head(exon_counts)

samdat_dif = filter(samdat, Differentiation == 'Diff')
exon_ct_dif = exon_counts[,rownames(samdat_dif)]
samdat_dif
# head(exon_ct_dif)

# dxd = DEXSeqDataSet(exon_ct_dif, samdat_dif,
#                     design = ~sample + exon + exon:Treatment,
#                     featureID = annot$exon_unique,
#                     groupID = annot$GeneID)
# dxd = estimateSizeFactors(dxd, locfunc = genefilter::shorth)
# dxd = estimateDispersions(dxd, BPPARAM = MulticoreParam(workers = 10))
# save(dxd, file = 'RData/dxd_dif.RData')
# plotDispEsts(dxd)
# dxd = testForDEU(dxd, BPPARAM = MulticoreParam(workers = 10))
# dxd = estimateExonFoldChanges(dxd, fitExpToVar = 'Treatment',
#                               BPPARAM = MulticoreParam(workers = 10))
# save(dxd, file = 'RData/dxd_dif_deu.RData')
load('RData/dxd_dif_deu.RData')
res_orig_dex_dif = DEXSeqResults(dxd)

# Identify genes with significanly different (padj < 0.05) exon usage in at 
# least one exon
# head(res_orig_dex_dif)
dim(res_orig_dex_dif)

# Remove the exons with insufficient data to analyze
res_dex_dif = (res_orig_dex_dif
       %>% data.frame()
       %>% filter(!is.na(padj)))
dim(res_dex_dif)

sum(res_dex_dif$padj < 0.05)
```

There are 1,520,722 exons analyzed. Of them, 1,191,946 had sufficient data for
analysis, and 1,237 had adjusted p-values less than 0.05.


Pull out the genes that have at least one DEU exon:

```{r}
sig_genes_dex_dif = (res_dex_dif
           %>% filter(padj < 0.05)
           %>% select(groupID)
           %>% unique())$groupID
length(sig_genes_dex_dif)

# Check if any of the problem genes have DEU
any(sig_genes_dex_dif %in% duplicate_geneIDs$gene_id)

sig_dex_dif = (res_dex_dif
                 %>% filter(padj < 0.05)
                 %>% select(-genomicData))
sig_10_dex_dif = (res_dex_dif
                    %>% filter(padj < 0.1)
                    %>% select(-genomicData))
# head(sig_dex_dif)
rownames(sig_dex_dif) = NULL
write.csv(sig_dex_dif, file = 'results/06_analysis/diff_deu_sig.csv')
write.csv(sig_10_dex_dif, file = 'results/06_analysis/diff_deu_0.01_sig.csv')
# write.csv(select(res_dex_dif, -genomicData), file = 'results/06_analysis/diff_deu_all.csv',
          # row.names = FALSE)
```

219 genes have at least one exon with DEU. None of the genes with duplicate IDs 
have DEU. You can download the significant results of DEXSeq
[here](./results/06_analysis/diff_deu_sig.csv').

#### Goseq

```{r}
all_genes = unique(res_orig_dex_dif$groupID)
length(all_genes)
genes_dex_dif = unique(res_dex_dif$groupID)
length(genes_dex_dif)
```

Of the 40,840 genes in the data set, 21,782 had sufficient coverage for analysis
in at least one exon.

```{r, warning=FALSE, message=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_dex_dif = genes_dex_dif %in% sig_genes_dex_dif
deg_v_dex_dif = as.numeric(deg_v_dex_dif)
names(deg_v_dex_dif) = genes_dex_dif

# remove the duplicate genes
deg_v_dex_dif = deg_v_dex_dif[!(genes_dex_dif %in% duplicate_geneIDs$gene_id)]
names(deg_v_dex_dif) = gene_id_v[names(deg_v_dex_dif)]
# head(deg_v_dex_dif)
any(is.na(names(deg_v_dex_dif)))

# Get the bias and length estimations for goseq
np_dex_dif = nullp(deg_v_dex_dif, 'mm39', 'refGene')

# Run goseq
gs_dex_dif = goseq(np_dex_dif, 'mm39', 'refGene')

# head(gs_dex_dif)
dim(gs_dex_dif)
gs_fixed_dex_dif = (gs_dex_dif
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_dex_dif)

# Count the significant categories
sum(sum(gs_fixed_dex_dif$padj < 0.05))

write.csv(filter(gs_fixed_dex_dif, padj < 0.05), 
          file = 'results/06_analysis/goseq_deu_diff.csv')
write.csv(gs_fixed_dex_dif, file = 'results/06_analysis/goseq_deu_diff_all.csv')
```

The 219 genes with significantly different exon usage produced 
over-representation of DEU in 0 GO categories (out of a total of 21,389 
possible categories, after BH FDR adjustment). You can download the file of
significant results [here](./results/06_analysis/goseq_deu_diff.csv') if you
want, but it's empty. The full results are 
[here](./results/06_analysis/goseq_deu_diff_all.csv).

### Guide Sequences

#### DEXSeq

I run this once and then keep it stored because it takes a long time to run

```{r}
# Run DEXSeq on guide sequence samples
head(exon_counts)
head(samdat)

# samdat_guide = filter(samdat, Treatment == 'KO')
# exon_ct_guide = exon_counts[,rownames(samdat_guide)]
# samdat_guide
# head(exon_ct_guide)
# 
# dxd = DEXSeqDataSet(exon_ct_guide, samdat_guide,
#                     design = ~sample + exon + exon:Differentiation,
#                     featureID = annot$exon_unique,
#                     groupID = annot$GeneID)
# dxd = estimateSizeFactors(dxd, locfunc = genefilter::shorth)
# dxd = estimateDispersions(dxd, BPPARAM = MulticoreParam(workers = 10))
# save(dxd, file = 'RData/dxd_guide.RData')
# plotDispEsts(dxd)
# dxd = testForDEU(dxd, BPPARAM = MulticoreParam(workers = 10))
# dxd = estimateExonFoldChanges(dxd, fitExpToVar = 'Differentiation',
#                               BPPARAM = MulticoreParam(workers = 8))
# save(dxd, file = 'RData/dxd_guide.RData')
load('RData/dxd_guide.RData')
res_orig_dex_guide = DEXSeqResults(dxd)

# Identify genes with significanly different (padj < 0.05) exon usage in at 
# least one exon
# head(res_orig_dex_guide)
dim(res_orig_dex_guide)

# Remove the exons with insufficient data to analyze
res_dex_guide = (res_orig_dex_guide
       %>% data.frame()
       %>% filter(!is.na(padj)))
dim(res_dex_guide)

sum(res_dex_guide$padj < 0.05)

sig_dex_guide = (res_dex_guide
                 %>% filter(padj < 0.05)
                 %>% select(-genomicData))
sig_10_dex_guide = (res_dex_guide
                 %>% filter(padj < 0.1)
                 %>% select(-genomicData))
# head(sig_dex_guide)
rownames(sig_dex_guide) = NULL
write.csv(sig_dex_guide, file = 'results/06_analysis/guide_deu_sig.csv')
write.csv(sig_10_dex_guide, file = 'results/06_analysis/guide_deu_0.01_sig.csv')
# write.csv(select(res_dex_guide, -genomicData), file = 'results/06_analysis/guide_deu_all.csv',
#             row.names = FALSE)
```

There are 1,520,722 exons analyzed. Of them, 1,164,641 had sufficient data for
analysis, and 6,789 had adjusted p-values less than 0.05. You can download the
significant results of DEXSeq analysis
[here](./results/06_analysis/guide_deu_sig.csv).


Pull out the genes that have at least one DEU exon:

```{r}
sig_genes_dex_guide = (res_dex_guide
           %>% filter(padj < 0.05)
           %>% select(groupID)
           %>% unique())$groupID
length(sig_genes_dex_guide)

# Check if any of the problem genes have DEU
any(sig_genes_dex_guide %in% duplicate_geneIDs$gene_id)
```

877 genes have at least one exon with DEU. None of the genes with duplicate IDs 
have DEU.

#### Goseq

```{r}
all_genes = unique(res_orig_dex_guide$groupID)
length(all_genes)
genes_dex_guide = unique(res_dex_guide$groupID)
length(genes_dex_guide)
```

Of the 40,840 genes in the data set, 20,537 had sufficient coverage for analysis
in at least one exon.

```{r, warning=FALSE, message=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_dex_guide = genes_dex_guide %in% sig_genes_dex_guide
deg_v_dex_guide = as.numeric(deg_v_dex_guide)
names(deg_v_dex_guide) = genes_dex_guide

# remove the duplicate genes
deg_v_dex_guide = deg_v_dex_guide[!(genes_dex_guide %in% duplicate_geneIDs$gene_id)]
names(deg_v_dex_guide) = gene_id_v[names(deg_v_dex_guide)]
# head(deg_v_dex_guide)
any(is.na(names(deg_v_dex_guide)))

# Get the bias and length estimations for goseq
np_dex_guide = nullp(deg_v_dex_guide, 'mm39', 'refGene')

# Run goseq
gs_dex_guide = goseq(np_dex_guide, 'mm39', 'refGene')

# head(gs_dex_guide)
dim(gs_dex_guide)
gs_fixed_dex_guide = (gs_dex_guide
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_dex_guide)

# Count the significant categories
sum(gs_fixed_dex_guide$padj < 0.05)

# Write the results to a file
write.csv(filter(gs_fixed_dex_guide, padj < 0.05), 
          file = 'results/06_analysis/goseq_deu_guide.csv')
write.csv(gs_fixed_dex_guide, file = 'results/06_analysis/goseq_deu_guide_all.csv')
```

The 877 genes with significantly different exon usage produced 
over-representation of DEU in 31 GO categories (out of a total of 21,389 
possible categories, after BH FDR adjustment). You can download the significant
results of the `goseq` analysis
[here](./results/06_analysis/goseq_deu_guide.csv), or the full results 
[here](./results/06_analysis/goseq_deu_guide_all.csv).


### Control Sequences

#### DEXSeq

I run this once and then keep it stored because it takes a long time to run

```{r}
# # Run DEXSeq on differentiated samples
# head(exon_counts)

samdat_ctl = filter(samdat, Treatment == 'WT')
exon_ct_ctl = exon_counts[,rownames(samdat_ctl)]
# samdat_ctl
# head(exon_ct_ctl)
# 
# dxd = DEXSeqDataSet(exon_ct_ctl, samdat_ctl,
#                     design = ~sample + exon + exon:Differentiation,
#                     featureID = annot$exon_unique,
#                     groupID = annot$GeneID)
# dxd = estimateSizeFactors(dxd, locfunc = genefilter::shorth)
# dxd = estimateDispersions(dxd, BPPARAM = MulticoreParam(workers = 6))
# save(dxd, file = 'RData/dxd_ctl.RData')
# plotDispEsts(dxd)
# dxd = testForDEU(dxd, BPPARAM = MulticoreParam(workers = 6))
# dxd = estimateExonFoldChanges(dxd, fitExpToVar = 'Differentiation',
#                               BPPARAM = MulticoreParam(workers = 6))
# save(dxd, file = 'RData/dxd_ctl_deu.RData')
load('RData/dxd_ctl_deu.RData')
res_orig_dex_ctl = DEXSeqResults(dxd)

# Identify genes with significanly different (padj < 0.05) exon usage in at 
# least one exon
# head(res_orig_dex_ctl)
dim(res_orig_dex_ctl)

# Remove the exons with insufficient data to analyze
res_dex_ctl = (res_orig_dex_ctl
       %>% data.frame()
       %>% filter(!is.na(padj)))
dim(res_dex_ctl)

sum(res_dex_ctl$padj < 0.05)
```

There are 1,520,722 exons analyzed. Of them, 1,154,143 had sufficient data for
analysis, and 27,967 had adjusted p-values less than 0.05.


Pull out the genes that have at least one DEU exon:

```{r}
sig_genes_dex_ctl = (res_dex_ctl
           %>% filter(padj < 0.05)
           %>% select(groupID)
           %>% unique())$groupID
length(sig_genes_dex_ctl)

# Check if any of the problem genes have DEU
any(sig_genes_dex_ctl %in% duplicate_geneIDs$gene_id)

sig_dex_ctl = (res_dex_ctl
                 %>% filter(padj < 0.05)
                 %>% select(-genomicData))
# head(sig_dex_ctl)
rownames(sig_dex_ctl) = NULL
write.csv(sig_dex_ctl, file = 'results/06_analysis/ctl_deu_sig.csv')
write.csv(select(res_dex_ctl, -genomicData), file = 'results/06_analysis/ctl_deu_all.csv',
          row.names = FALSE)
```

3,135 genes have at least one exon with DEU. None of the genes with duplicate IDs 
have DEU. You can download the significant results of DEXSeq
[here](./results/06_analysis/ctl_deu_sig.csv').

#### Goseq

```{r}
all_genes = unique(res_orig_dex_ctl$groupID)
length(all_genes)
genes_dex_ctl = unique(res_dex_ctl$groupID)
length(genes_dex_ctl)
```

Of the 40,840 genes in the data set, 20,299 had sufficient coverage for analysis
in at least one exon.

```{r, warning=FALSE, message=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_dex_ctl = genes_dex_ctl %in% sig_genes_dex_ctl
deg_v_dex_ctl = as.numeric(deg_v_dex_ctl)
names(deg_v_dex_ctl) = genes_dex_ctl

# remove the duplicate genes
deg_v_dex_ctl = deg_v_dex_ctl[!(genes_dex_ctl %in% duplicate_geneIDs$gene_id)]
names(deg_v_dex_ctl) = gene_id_v[names(deg_v_dex_ctl)]
# head(deg_v_dex_ctl)
any(is.na(names(deg_v_dex_ctl)))

# Get the bias and length estimations for goseq
np_dex_ctl = nullp(deg_v_dex_ctl, 'mm39', 'refGene')

# Run goseq
gs_dex_ctl = goseq(np_dex_ctl, 'mm39', 'refGene')

# head(gs_dex_ctl)
dim(gs_dex_ctl)
gs_fixed_dex_ctl = (gs_dex_ctl
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_dex_ctl)

# Count the significant categories
sum(sum(gs_fixed_dex_ctl$padj < 0.05))

write.csv(filter(gs_fixed_dex_ctl, padj < 0.05), 
          file = 'results/06_analysis/goseq_deu_ctl.csv')
write.csv(gs_fixed_dex_ctl, file = 'results/06_analysis/goseq_deu_ctl_all.csv')
```

The 3,135 genes with significantly different exon usage produced 
over-representation of DEU in 110 GO categories (out of a total of 21,613 
possible categories, after BH FDR adjustment). You can download the file of
significant results [here](./results/06_analysis/goseq_deu_ctl.csv') if you
want, but it's empty. The full results are 
[here](./results/06_analysis/goseq_deu_ctl_all.csv).


## Differential Expression

```{r}
# names(fc_lst_gene[[1]])
# dim(fc_lst_gene[[1]]$counts)

gene_counts = matrix(nrow = nrow(fc_lst_gene[[1]]$counts),
                     ncol = length(fc_lst_gene))
samps = names(fc_lst_gene)
samps = gsub('-','_',samps)
colnames(gene_counts) = samps
# head(gene_counts)
names(fc_lst_gene) = samps

for(samp in samps){
    gene_counts[,samp] = fc_lst_gene[[samp]]$counts
}

# head(gene_counts)
```

### Interaction Model

#### DESeq

```{r, warning=FALSE, message=FALSE}
samdat
head(gene_counts)

dds = DESeqDataSetFromMatrix(gene_counts, samdat, ~Treatment*Differentiation)
dds = DESeq(dds)
resultsNames(dds)
```

##### Naive Cell Results

```{r}
resultsNames(dds)
res_undif = lfcShrink(dds, contrast = c('Treatment', 'KO', 'WT'), 
                      type = 'ashr')
res_undif_df = data.frame(res_undif)
rownames(res_undif_df) = fc_lst_gene[[1]]$annotation$GeneID
head(res_undif_df)
dim(res_undif_df)
dim(gene_counts)
res_undif_df = cbind(res_undif_df, gene_counts)
head(res_undif_df)
sum(!is.na(res_undif_df$padj))
sum(res_undif_df$padj < 0.05, na.rm = TRUE)

sig_undif = (res_undif_df
             %>% filter(padj < 0.05))
write.csv(sig_undif, file = 'results/06_analysis/naive_des_sig.csv')
write.csv(res_undif_df, file = 'results/06_analysis/naive_des_all.csv')
```

Of the 40,840 genes analysed, 20,969 had sufficient coverage for analysis. Of
those, 804 were significantly differentially expressed between the KO and WT
treatments within _undifferentiated_ cells.

##### Differentiated Cell Results

```{r}
resultsNames(dds)
res_dif = lfcShrink(dds, contrast = list(c('Treatment_KO_vs_WT', 
                                           'TreatmentKO.DifferentiationDiff')),
                      type = 'ashr')
res_dif_df = data.frame(res_dif)
rownames(res_dif_df) = fc_lst_gene[[1]]$annotation$GeneID
head(res_dif_df)
dim(res_dif_df)
dim(gene_counts)
res_dif_df = cbind(res_dif_df, gene_counts)
head(res_dif_df)
sum(!is.na(res_dif_df$padj))
sum(res_dif_df$padj < 0.05, na.rm = TRUE)

sig_dif = (res_dif_df
             %>% filter(padj < 0.05))
write.csv(sig_dif, file = 'results/06_analysis/diff_des_sig.csv')
write.csv(res_dif_df, file = 'results/06_analysis/diff_des_all.csv')
```

Of the 40,840 genes analysed, 20,395 had sufficient coverage for analysis. Of
those, 1,173 were significantly differentially expressed between the KO and WT
treatments within _differentiated_ cells.

##### Control Cell Results

```{r}
resultsNames(dds)
res_ctl = lfcShrink(dds, contrast = c('Differentiation', 'Diff', 'Naive'), 
                      type = 'ashr')
res_ctl_df = data.frame(res_ctl)
rownames(res_ctl_df) = fc_lst_gene[[1]]$annotation$GeneID
head(res_ctl_df)
dim(res_ctl_df)
dim(gene_counts)
res_ctl_df = cbind(res_ctl_df, gene_counts)
head(res_ctl_df)
sum(!is.na(res_ctl_df$padj))
sum(res_ctl_df$padj < 0.05, na.rm = TRUE)

sig_ctl = (res_ctl_df
             %>% filter(padj < 0.05))
write.csv(sig_ctl, file = 'results/06_analysis/ctl_des_sig.csv')
write.csv(res_ctl_df, file = 'results/06_analysis/ctl_des_all.csv')
```

Of the 40,840 genes analysed, 20,969 had sufficient coverage for analysis. Of
those, 6,112 were significantly differentially expressed between the Naive and 
Differentiated cells within the _scramble control WT_ treatment.

##### Guide Cell Results

```{r}
resultsNames(dds)
res_guide = lfcShrink(dds, contrast = list(c('Differentiation_Diff_vs_Naive',
                                           'TreatmentKO.DifferentiationDiff')),
                      type = 'ashr')
res_guide_df = data.frame(res_guide)
rownames(res_guide_df) = fc_lst_gene[[1]]$annotation$GeneID
head(res_guide_df)
dim(res_guide_df)
dim(gene_counts)
res_guide_df = cbind(res_guide_df, gene_counts)
head(res_guide_df)
sum(!is.na(res_guide_df$padj))
sum(res_guide_df$padj < 0.05, na.rm = TRUE)

sig_guide = (res_guide_df
             %>% filter(padj < 0.05))
write.csv(sig_guide, file = 'results/06_analysis/guide_des_sig.csv')
write.csv(res_guide_df, file = 'results/06_analysis/guide_des_all.csv')
```

Of the 40,840 genes analysed, 20,969 had sufficient coverage for analysis. Of
those, 4,099 were significantly differentially expressed between the Naive and
Differentiated cells within the _guide KO_ treatment.

##### Interaction Term Result

```{r}
resultsNames(dds)
res_des = lfcShrink(dds, contrast = list(c('TreatmentKO.DifferentiationDiff')),
                    type = 'ashr')
res_df_des = data.frame(res_des)
rownames(res_df_des) = fc_lst_gene[[1]]$annotation$GeneID
head(res_df_des)
dim(res_df_des)
dim(gene_counts)
res_df_des = cbind(res_df_des, gene_counts)
head(res_df_des)
sum(!is.na(res_df_des$padj))
sum(res_df_des$padj < 0.05, na.rm = TRUE)

sig_des = (res_df_des
                 %>% filter(padj < 0.05))
# head(sig_des_undif)
write.csv(sig_des, file = 'results/06_analysis/interaction_des_sig.csv')
write.csv(res_df_des, file = 'results/06_analysis/interaction_des_all.csv')
```

Of the 40,840 genes analyzed, 30,155 had sufficient coverage for analysis, and 
none were significant in the interaction term, which is quite odd, but likely
reflects the noisiness of the data more than anything else.

#### Goseq

```{r}
all_genes_des_undif = rownames(res_df_des_undif)
# length(all_genes_des_undif)
genes = (res_df_des_undif
         %>% filter(!is.na(padj))
         %>% rownames())
# length(genes)

sig_genes_des_undif = (res_df_des_undif
             %>% filter(padj < 0.05)
             %>% rownames())

```


```{r, message=FALSE, warning=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_des_undif = genes %in% sig_genes_des_undif
deg_v_des_undif = as.numeric(deg_v_des_undif)
names(deg_v_des_undif) = genes

# remove the duplicate genes
deg_v_des_undif = deg_v_des_undif[!(genes %in% duplicate_geneIDs$gene_id)]
names(deg_v_des_undif) = gene_id_v[names(deg_v_des_undif)]
# head(deg_v_des_undif)
any(is.na(names(deg_v_des_undif)))

# Get the bias and length estimations for goseq
np_des_undif = nullp(deg_v_des_undif, 'mm39', 'refGene')

# Run goseq
gs_des_undif = goseq(np_des_undif, 'mm39', 'refGene')

# head(gs_des_undif)
dim(gs_des_undif)
gs_fixed_des_undif = (gs_des_undif
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_des_undif)

# Count the significant categories
sum(gs_fixed_des_undif$padj < 0.05)

# head(gs_fixed_des_undif)
write.csv(filter(gs_fixed_des_undif, padj < 0.05), 
          file = 'results/06_analysis/goseq_de_naive.csv')
write.csv(gs_fixed_des_undif, file = 'results/06_analysis/goseq_de_naive_all.csv')
```

The 1,521 genes that were significantly differentially expressed led to 
over-representation of difference in 219 GO categories. You can download the
significant `goseq` results [here](./results/06_analysis/goseq_de_naive.csv'),
or the full results [here](./results/06_analysis/goseq_de_naive_all.csv).

### Naive Cells

#### DESeq

```{r, warning=FALSE, message=FALSE}
samdat_undif
gene_ct_undif = gene_counts[,rownames(samdat_undif)]

dds_undif = DESeqDataSetFromMatrix(gene_ct_undif, samdat_undif, ~Treatment)
dds_undif = DESeq(dds_undif)
resultsNames(dds_undif)

res_des_undif = lfcShrink(dds_undif, coef = 'Treatment_KO_vs_WT')
res_df_des_undif = data.frame(res_des_undif)
rownames(res_df_des_undif) = fc_lst_gene[[1]]$annotation$GeneID
head(res_df_des_undif)
dim(res_df_des_undif)
dim(gene_ct_undif)
res_df_des_undif = cbind(res_df_des_undif, gene_ct_undif)
head(res_df_des_undif)
sum(!is.na(res_df_des_undif$padj))
sum(res_df_des_undif$padj < 0.05, na.rm = TRUE)

sig_des_undif = (res_df_des_undif
                 %>% filter(padj < 0.05))
# head(sig_des_undif)
write.csv(sig_des_undif, file = 'results/06_analysis/naive_des_sig.csv')
write.csv(res_df_des_undif, file = 'results/06_analysis/naive_des_all.csv')
```

Of the 40,840 genes analyzed, 19,448 had sufficient coverage for analysis, and 
1,521 were significantly differentially expressed. You can download the
significantly differentially expressed gene
[here](./results/06_analysis/naive_des_sig.csv), or the full results 
[here](./results/06_analysis/naive_des_all.csv).

#### Goseq

```{r}
all_genes_des_undif = rownames(res_df_des_undif)
# length(all_genes_des_undif)
genes = (res_df_des_undif
         %>% filter(!is.na(padj))
         %>% rownames())
# length(genes)

sig_genes_des_undif = (res_df_des_undif
             %>% filter(padj < 0.05)
             %>% rownames())

```


```{r, message=FALSE, warning=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_des_undif = genes %in% sig_genes_des_undif
deg_v_des_undif = as.numeric(deg_v_des_undif)
names(deg_v_des_undif) = genes

# remove the duplicate genes
deg_v_des_undif = deg_v_des_undif[!(genes %in% duplicate_geneIDs$gene_id)]
names(deg_v_des_undif) = gene_id_v[names(deg_v_des_undif)]
# head(deg_v_des_undif)
any(is.na(names(deg_v_des_undif)))

# Get the bias and length estimations for goseq
np_des_undif = nullp(deg_v_des_undif, 'mm39', 'refGene')

# Run goseq
gs_des_undif = goseq(np_des_undif, 'mm39', 'refGene')

# head(gs_des_undif)
dim(gs_des_undif)
gs_fixed_des_undif = (gs_des_undif
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_des_undif)

# Count the significant categories
sum(gs_fixed_des_undif$padj < 0.05)

# head(gs_fixed_des_undif)
write.csv(filter(gs_fixed_des_undif, padj < 0.05), 
          file = 'results/06_analysis/goseq_de_naive.csv')
write.csv(gs_fixed_des_undif, file = 'results/06_analysis/goseq_de_naive_all.csv')
```

The 1,521 genes that were significantly differentially expressed led to 
over-representation of difference in 219 GO categories. You can download the
significant `goseq` results [here](./results/06_analysis/goseq_de_naive.csv'),
or the full results [here](./results/06_analysis/goseq_de_naive_all.csv).

### Differentiated Cells

#### DESeq

```{r, message=FALSE, warning=FALSE}
samdat_dif
gene_ct_dif = gene_counts[,rownames(samdat_dif)]

dds_dif = DESeqDataSetFromMatrix(gene_ct_dif, samdat_dif, ~Treatment)
dds_dif = DESeq(dds_dif)
resultsNames(dds_dif)

res_des_dif = lfcShrink(dds_dif, coef = 'Treatment_KO_vs_WT')
res_df_des_dif = data.frame(res_des_dif)
rownames(res_df_des_dif) = fc_lst_gene[[1]]$annotation$GeneID
head(res_df_des_dif)
dim(res_df_des_dif)
dim(gene_ct_dif)
head(gene_ct_dif)
res_df_des_dif = cbind(res_df_des_dif, gene_ct_dif)
sum(!is.na(res_df_des_dif$padj))
sum(res_df_des_dif$padj < 0.05, na.rm = TRUE)

sig_des_dif = (res_df_des_dif
                 %>% filter(padj < 0.05))
# head(sig_des_dif)
write.csv(sig_des_dif, file = 'results/06_analysis/diff_des_sig.csv')
write.csv(res_df_des_dif, file = 'results/06_analysis/diff_des_all.csv')
```

Of the 40,840 genes analyzed, 20,403 had sufficient coverage for analysis, and 
531 were significantly differentially expressed. You can download the
significantly differentially expressed genes
[here](./results/06_analysis/diff_des_sig.csv), or the full results
[here](./results/06_analysis/diff_des_all.csv).

#### Goseq

```{r, warning=FALSE,message=FALSE}
all_genes = rownames(res_df_des_dif)
# length(all_genes)
genes_des_dif = (res_df_des_dif
         %>% filter(!is.na(padj))
         %>% rownames())
# length(genes_des_dif)

sig_genes_des_dif = (res_df_des_dif
             %>% filter(padj < 0.05)
             %>% rownames())
```


```{r, message=FALSE, warning=FALSE}
# Make a binary vector of all genes, 1 for significant, 0 for non
deg_v_des_dif = genes_des_dif %in% sig_genes_des_dif
deg_v_des_dif = as.numeric(deg_v_des_dif)
names(deg_v_des_dif) = genes_des_dif

# remove the duplicate genes
deg_v_des_dif = deg_v_des_dif[!(genes_des_dif %in% duplicate_geneIDs$gene_id)]
names(deg_v_des_dif) = gene_id_v[names(deg_v_des_dif)]
# head(deg_v_des_dif)
any(is.na(names(deg_v_des_dif)))

# Get the bias and length estimations for goseq
np_des_dif = nullp(deg_v_des_dif, 'mm39', 'refGene')

# Run goseq
gs_des_dif = goseq(np_des_dif, 'mm39', 'refGene')

# head(gs_des_dif)
dim(gs_des_dif)
gs_fixed_des_dif = (gs_des_dif
            %>% mutate(padj = p.adjust(over_represented_pvalue, 'BH')))
# head(gs_fixed_des_dif)

# Count the significant categories
sum(gs_fixed_des_dif$padj < 0.05)

write.csv(filter(gs_fixed_des_dif, padj < 0.05), 
          file = 'results/06_analysis/goseq_de_diff.csv')
write.csv(gs_fixed_des_dif, file = 'results/06_analysis/goseq_de_diff_all.csv')
```

The 513 differentially expressed genes led to over-representation of difference
in 138 GO categories. You can download the significant `goseq` results
[here](./results/06_analysis/goseq_de_diff.csv), or the full results 
[here](./results/06_analysis/goseq_de_diff_all.csv)


